pkg/kubelet/status/state/state_checkpoint_test.go

```diff
@@ -0,0 +1,108 @@
+package state
+
+import (
+	"fmt"
+	"testing"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/kubernetes/pkg/kubelet/checkpointmanager"
+)
+
+func Test_stateCheckpoint_storeState(t *testing.T) {
+	cache := NewStateMemory()
+	checkpointManager, _ := checkpointmanager.NewCheckpointManager(".")
+	checkpointName := "pod_state_checkpoint"
+	sc := &stateCheckpoint{
+		cache:             cache,
+		checkpointManager: checkpointManager,
+		checkpointName:    checkpointName,
+	}
+	var err error
+	type args struct {
+		podResourceAllocation PodResourceAllocation
+	}
+
+	tests := []struct {
+		name string
+		args args
+	}{}
+	suffix := []string{"Ki", "Mi", "Gi", "Ti", "Pi", "Ei", "n", "u", "m", "k", "M", "G", "T", "P", "E", ""}
+	factor := []string{"1", "0.1", "0.03", "10", "100", "512", "1000", "1024", "700", "10000"}
+	for j, fact := range factor {
+		for i, suf := range suffix {
+			tests = append(tests, struct {
+				name string
+				args args
+			}{
+				name: fmt.Sprintf("format %d - cpu %s%s", j*len(suffix)+i, fact, suf),
+				args: args{
+					podResourceAllocation: PodResourceAllocation{
+						"pod1": {
+							"container1": {
+								v1.ResourceCPU:    resource.MustParse(fmt.Sprintf("%s%s", fact, suf)),
+								v1.ResourceMemory: resource.MustParse(fmt.Sprintf("%s%s", fact, suf)),
+							},
+						},
+					},
+				},
+			})
+		}
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// reset environment
+			err = sc.cache.ClearState()
+			if err != nil {
+				t.Errorf("failed to clear state: %v", err)
+			}
+			// remove file ./stateDir/pod_state_checkpoint
+			err = checkpointManager.RemoveCheckpoint(checkpointName)
+			if err != nil {
+				t.Errorf("failed to remove checkpoint: %v", err)
+			}
+			err = sc.cache.SetPodResourceAllocation(tt.args.podResourceAllocation)
+			if err != nil {
+				t.Errorf("failed to set pod resource allocation: %v", err)
+			}
+			// store state
+			err = sc.storeState()
+			if err != nil {
+				t.Errorf("failed to store state: %v", err)
+			}
+			// reset cache
+			originCache := sc.cache
```

Comment:
I believe you should deep copy this, otherwise any change to underlying structures of sc.cache will change it, e.g. it will be cleaned by `sc.cache.ClearState()` below.

Reply:
Fixed. Thx